### PR TITLE
replace length(x) == 0 with isempty(x), similarly for !isempty

### PR DESCRIPTION
**Pull request by stevengj**
I noticed that in some places we were doing `length(string) > 0` to check whether it was not empty, which is O(n), whereas `!isempty(string)` is O(1).  (It probably wasn't being done in performance-critical cases, but it bugged me every time I saw it.)

Rather than find the places where `length(x)` is being compared to `0` specifically for strings, it was a lot easier to just globally search and replace this with `isempty(x)` or `!isempty(x)` everywhere.   It certainly shouldn't _hurt_ performance, and seemed like better style to use `isempty` to check for empty collections.

### CODE REVIEW

**Comment by stevengj** [comment]
Hmm, just noticed a breakage because `CompositeException` defined `length` but not `start` and `done`, hence `isempty` broke.   Since this object should clearly at least implement the iterator protocol, I added that.

Probably it should also have `getindex`, and probably it should be an array of `Exception` rather than of `Any` (but then we need `CapturedException <: Exception`, which should probably be the case anyway), and probably it should be documented better, but that seemed to something for another PR.

**Comment by tkelman** [comment]
+1

File: base/test.jl [hunk 1]
```diff
@@ -505,7 +505,7 @@ By default the `@testset` macro will return the testset object itself, though
 this behavior can be customized in other testset types.
 """
 macro testset(args...)
-    length(args) > 0 || error("No arguments to @testset")
+    !isempty(args) || error("No arguments to @testset")
```
Comment (IainNZ) [inline base/test.jl]: This is a nasty double negative now, could have just been `isempty(args) &&` surely?

Comment (stevengj) [inline base/test.jl]: Yes, I wasn't that clever with my regex.

**Pull request merged**